marlowe-plutus/charli3/Language/Marlowe/Scripts/Charli3.hs

```diff
@@ -0,0 +1,217 @@
+  -> V1.CurrencySymbol
+  -- ^ The policy ID for the Charli3 oracle feed.
+  -> V1.TokenName
+  -- ^ The token name from the Charli3 oracle feed.
+  -> V1.ChoiceName
+  -- ^ Name of the oracle choice in Marlowe.
+  -> (PV2.PubKeyHash, V1.TokenName)
+  -- ^ Datum should be the public key hash of the oracle operator and a thread token name.
+  -> Bool
+  -- ^ The redeemer should be a boolean indicating whether the script continues running.
+  -> ScriptContext
+  -- ^ The script context.
+  -> Bool
+mkValidator _ _ _ _ (pkh, _) False ScriptContext{scriptContextTxInfo = txInfo} = txSignedBy txInfo pkh
+mkValidator
+  marloweValidatorHash
+  charli3CurrencySymbol
+  charli3TokenName
+  charli3ChoiceName
+  (_, threadTokenName)
+  True
+  ScriptContext{scriptContextTxInfo = txInfo@TxInfo{..}, scriptContextPurpose} = do
+    let ownInput =
+          case scriptContextPurpose of
+            Spending txOutRef ->
+              case find (\TxInInfo{txInInfoOutRef} -> txInInfoOutRef == txOutRef) txInfoInputs of
+                Just input -> input
+                Nothing -> traceError "C3a" -- Impossible, script input not found.
+            _ -> traceError "C3b" -- Not a spending script.
+        TxInInfo{txInInfoResolved = TxOut{txOutAddress = ownAddress, txOutValue = ownValue, txOutDatum = ownDatum}} = ownInput
+
+        ownOutputOk =
+          case filter (\TxOut{txOutAddress} -> ownAddress == txOutAddress) txInfoOutputs of
+            [TxOut{txOutValue, txOutDatum}] -> txOutValue == ownValue && txOutDatum == ownDatum
+            _ -> traceError "C3c" -- Script value or datum changed.
+        PV2.Datum charli3Datum =
+          case filter
+            (\TxInInfo{txInInfoResolved = TxOut{txOutValue}} -> valueOf txOutValue charli3CurrencySymbol charli3TokenName > 0)
+            txInfoReferenceInputs of
+            [TxInInfo{txInInfoResolved = TxOut{txOutDatum = OutputDatum datum}}] -> datum
+            [TxInInfo{txInInfoResolved = TxOut{txOutDatum = OutputDatumHash hash}}] ->
+              case findDatum hash txInfo of
+                Just datum -> datum
+                Nothing -> traceError "C3d" -- Charli3 datum not found.
+            _ -> traceError "C3e" -- Charli3 oracle feed not found.
+        charli3Oracle =
+          case C3.getPriceDatas . C3.OracleFeed $ charli3Datum of
+            [pd] -> C3.getPrice . C3.getPriceMap $ pd
+            _ -> traceError "C3f" -- Precisely one Charli3 price value not found.
+        (currencySymbol, roleName) = do
+          let valuesList = AssocMap.toList $ getValue ownValue
+          case valuesList of
+            [(possibleAdaSymbol, _), (currencySymbol, AssocMap.toList -> [(roleName, _)])]
+              | possibleAdaSymbol == adaSymbol -> (currencySymbol, roleName)
+            [(currencySymbol, AssocMap.toList -> [(roleName, _)]), _] -> (currencySymbol, roleName)
+            _ -> traceError "C3g" -- Precisely one role token not found.
```

Comment:
I wonder whether it would make the validator slightly smaller to remove the ada symbol checking before matching, so we would have two case blocks, one for the ada bit, and one for the role extraction. Because we are using the same pattern twice. It is splitting hairs, I know, but I was bothered by the lack of symmetry in the check for ada and I gave it some thought

Reply:
@palas I'm not sure if I understand the suggestion. Using `''` inside the `case` pattern match doesn't work because of some equality operator picking issue as far as I recall.

What we could do probably here is make `AssocMap.toList` to be called lazily in body of a particular case because Plutus is eager and evaluates this right away if if the check fails.

Reply:
> make the validator slightly smaller
Since this is an experimental un-audited validator subject to future revision, I don't think we should undertake an effort to minimize validator size. If this validator goes into production, we should do that, plus add a test suite.

Reply:
Just for the record, what I meant was doing one case for removing the ada output from the list, and then pattern matching on the roleName on a separate statement (once you know which one is the right one).

Reply:
> one case for removing the ada output (..)

@palas Could you elaborate how it will look like? Would it involve function call and internal pattern match?
This elaborated pattern matching (and not use of convenient functions) was actually done on purpose here to minimize the overhead of function calls and internal pattern matches (evaluation actually ~= memory usage in Plutus core). `case` pattern matching seems to be really fast and cheap operation in Plutus-Tx.